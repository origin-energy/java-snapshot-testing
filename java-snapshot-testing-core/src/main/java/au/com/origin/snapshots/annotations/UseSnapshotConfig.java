package au.com.origin.snapshots.annotations;

import au.com.origin.snapshots.config.SnapshotConfig;

import java.lang.annotation.*;

@Documented
@Target({ElementType.TYPE})
@Inherited
@Retention(RetentionPolicy.RUNTIME)
public @interface UseSnapshotConfig {
  Class<? extends SnapshotConfig> value();
}
