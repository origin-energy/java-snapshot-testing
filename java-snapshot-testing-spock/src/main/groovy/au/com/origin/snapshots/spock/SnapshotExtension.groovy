package au.com.origin.snapshots.spock

import au.com.origin.snapshots.SnapshotVerifier
import au.com.origin.snapshots.config.PropertyResolvingSnapshotConfig
import au.com.origin.snapshots.config.SnapshotConfig
import au.com.origin.snapshots.config.SnapshotConfigInjector
import org.spockframework.runtime.extension.AbstractAnnotationDrivenExtension
import org.spockframework.runtime.model.SpecInfo

class SnapshotExtension extends AbstractAnnotationDrivenExtension<EnableSnapshots> implements SnapshotConfigInjector {

    SnapshotVerifier snapshotVerifier;

    void visitSpecAnnotation(EnableSnapshots annotation, SpecInfo spec) {
        this.snapshotVerifier = new SnapshotVerifier(getSnapshotConfig(), spec.reflection, false)
    }

    void visitSpec(SpecInfo spec) {
        spec.allFeatures.featureMethod*.addInterceptor(new SnapshotMethodInterceptor(snapshotVerifier))
    }



    @Override
    SnapshotConfig getSnapshotConfig() {
        return new PropertyResolvingSnapshotConfig()
    }
}