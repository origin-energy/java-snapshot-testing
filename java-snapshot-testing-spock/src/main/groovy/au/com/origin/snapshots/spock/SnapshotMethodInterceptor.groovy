package au.com.origin.snapshots.spock

import au.com.origin.snapshots.Expect
import au.com.origin.snapshots.ReflectionUtils
import au.com.origin.snapshots.SnapshotVerifier
import au.com.origin.snapshots.logging.LoggingHelper
import org.slf4j.LoggerFactory
import org.spockframework.runtime.extension.AbstractMethodInterceptor
import org.spockframework.runtime.extension.IMethodInvocation

import java.lang.reflect.Method

// Based on this issue: https://github.com/spockframework/spock/issues/652
class SnapshotMethodInterceptor extends AbstractMethodInterceptor {
    private log = LoggerFactory.getLogger( SnapshotMethodInterceptor.class )
    private final SnapshotVerifier snapshotVerifier;

    SnapshotMethodInterceptor(SnapshotVerifier snapshotVerifier) {
        this.snapshotVerifier = snapshotVerifier
    }

    @Override
    void interceptFeatureMethod(IMethodInvocation invocation) throws Throwable {
        updateInstanceVariable(invocation.instance, invocation.feature.featureMethod.reflection)

        def parameterCount = invocation.method.reflection.parameterCount
        if (parameterCount > invocation.arguments.length) {
            def newArguments = new Object[parameterCount]
            System.arraycopy invocation.arguments, 0, newArguments, 0, invocation.arguments.length
            invocation.arguments = newArguments
        }
        invocation.method.reflection.parameterTypes.eachWithIndex { type, i ->
            if (Expect.class == type) {
                LoggingHelper.deprecatedV5(log, "Injecting 'Expect' via method a argument is no longer recommended. Consider using instance variable injection instead.")
                invocation.arguments[i] = new Expect(snapshotVerifier, invocation.feature.featureMethod.reflection)
            }
        }
        invocation.proceed()
    }

    private void updateInstanceVariable(Object testInstance, Method testMethod) {
        ReflectionUtils.findFieldByPredicate(testInstance.class, { field -> field.getType() == Expect.class })
            .ifPresent({ field ->
                Expect expect = Expect.of(snapshotVerifier, testMethod);
                ReflectionUtils.makeAccessible(field);
                try {
                    field.set(testInstance, expect);
                } catch (IllegalAccessException e) {
                    throw new RuntimeException(e);
                }
            });
    }

    @Override
    void interceptCleanupSpecMethod(IMethodInvocation invocation) throws Throwable {
        this.snapshotVerifier.validateSnapshots();
    }
}
