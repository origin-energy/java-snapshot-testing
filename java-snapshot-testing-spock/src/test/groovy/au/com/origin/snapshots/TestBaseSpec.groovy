package au.com.origin.snapshots

import au.com.origin.snapshots.annotations.SnapshotName
import au.com.origin.snapshots.spock.EnableSnapshots

@EnableSnapshots
class TestBaseSpec extends SpecificationBase {

    @SnapshotName("Should use extension")
    def "Should use extension"() {
        when:
        expect.toMatchSnapshot("Hello World")

        then:
        true
    }

}
