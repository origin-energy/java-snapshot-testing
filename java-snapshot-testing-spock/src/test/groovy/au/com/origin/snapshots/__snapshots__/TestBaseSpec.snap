Should use extension=[
Hello World
]